sdk/core/System.ClientModel/src/Message/MultipartContent.cs

```diff
@@ -0,0 +1,295 @@
+
+    /// <summary>
+    /// Adds a new <see cref="BinaryContent"/> instance to the collection of <see cref="BinaryContent"/> objects that get
+    /// formatted into multipart/* content.
+    /// </summary>
+    /// <param name="content">The <see cref="BinaryContent"/> to add to the collection.</param>
+    /// <param name="headers">The headers to add to this <see cref="BinaryContent"/>'s section.</param>
+    public void Add(BinaryContent content, Dictionary<string, string> headers)
+    {
+        if (content == null)
+        {
+            throw new ArgumentNullException(nameof(content));
+        }
+
+        if (headers == null)
+        {
+            throw new ArgumentNullException(nameof(headers));
+        }
+
+        var subpart = new MultipartContentSubpart(content, headers);
+
+        _subparts.Add(new MultipartContentSubpart(content, headers));
+    }
+
+    private static void ValidateBoundary(string boundary)
+    {
+        if (string.IsNullOrWhiteSpace(boundary))
+        {
+            throw new ArgumentException("Value cannot be null or empty.", boundary);
+        }
+
+        // RFC 2046 Section 5.1.1
+
+        if (boundary.Length > 70)
+        {
+            throw new ArgumentOutOfRangeException(nameof(boundary), boundary, $"The field cannot be longer than {70} characters.");
+        }
+
+        if (boundary.EndsWith(" ", StringComparison.InvariantCultureIgnoreCase))
+        {
+            throw new ArgumentException($"The format of value '{boundary}' is invalid. '{boundary}' cannot end in a space", nameof(boundary));
+        }
+
+        foreach (char ch in boundary)
+        {
+            const string AllowedMarks = @"'()+_,-./:=? ";
+            if (!(('0' <= ch && ch <= '9') || // Digit.
+                    ('a' <= ch && ch <= 'z') || // alpha.
+                    ('A' <= ch && ch <= 'Z') || // ALPHA.
+                    AllowedMarks.Contains(char.ToString(ch)))) // Marks.
+            {
+                throw new ArgumentException($"The format of value '{boundary}' is invalid.", nameof(boundary));
+            }
+        }
+    }
+
+    /// <summary>
+    /// Releases the unmanaged resources and disposes of the managed resources used by the <see cref="MultipartContent"/>.
+    /// </summary>
+    public void Dispose()
+    {
+    }
+
+    #region Subpart
+
+    /// <summary>
+    /// A helper class representing a subpart of a multipart content.
+    /// </summary>
+    private class MultipartContentSubpart
+    {
+        /// <summary>
+        /// The <see cref="BinaryContent"/> that represents the content of this subpart.
+        /// </summary>
+        public BinaryContent Content { get; }
+
+        /// <summary>
+        /// The headers for this subpart.
+        /// </summary>
+        public Dictionary<string, string> Headers { get; }
+
+        public MultipartContentSubpart(BinaryContent content, Dictionary<string, string> headers)
```

Comment:
Can we define this a public class? we may need this when handle multipart/form-data response.

Reply:
In general, it's best not to expose public APIs until we have a clear end-user scenario that requires them.  @m-redding is correct that we can make this public later if needed, but if we made it public now and discovered the design of the APIs didn't address user requirements, we would be stuck with them and unable to change them later.